File: check_syntax.js
const fs = require('fs');

try {
    const html = fs.readFileSync('index.html', 'utf8');
    const startIdx = html.indexOf('<script>');
    const endIdx = html.lastIndexOf('</script>');
    
    if (startIdx === -1 || endIdx === -1) {
        console.log("No script tag found!");
        process.exit(1);
    }
    
    const js = html.substring(startIdx + 8, endIdx);
    fs.writeFileSync('temp.js', js, 'utf8');
    console.log("JS extracted to temp.js");
    
    // Check brackets balance
    const stack = [];
    const lines = js.split('\n');
    let hasMismatches = false;
    
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        for (let j = 0; j < line.length; j++) {
            const char = line[j];
            if (char === '{' || char === '[' || char === '(') {
                stack.push({ char, line: i + 1 });
            } else if (char === '}' || char === ']' || char === ')') {
                if (stack.length === 0) {
                    console.log(`Extra closing char '${char}' at line ${i + 1}`);
                    hasMismatches = true;
                } else {
                    const top = stack.pop();
                    if ((char === '}' && top.char !== '{') ||
                        (char === ']' && top.char !== '[') ||
                        (char === ')' && top.char !== '(')) {
                        console.log(`Mismatched '${top.char}' (line ${top.line}) and '${char}' (line ${i + 1})`);
                        hasMismatches = true;
                    }
                }
            }
        }
    }
    
    if (stack.length > 0) {
        console.log("Unclosed brackets:");
        stack.slice(0, 10).forEach(item => {
            console.log(`  '${item.char}' at line ${item.line}`);
        });
        hasMismatches = true;
    }
    
    if (!hasMismatches) {
        console.log("Brackets balance: OK");
    } else {
        console.log("Brackets balance: FAILED");
    }
    
    // Try evaluating using new Function (checks syntax)
    try {
        new Function(js);
        console.log("Syntax validation: OK");
    } catch(e) {
        console.log("Syntax validation: FAILED");
        console.error(e);
    }
    
} catch(e) {
    console.log("Failed to run syntax checker:", e);
}

File: check_syntax.py
import re
import subprocess
import os

try:
    with open('index.html', 'r', encoding='utf-8') as f:
        html = f.read()
    
    scripts = re.findall(r'<script>(.*?)</script>', html, re.DOTALL)
    if not scripts:
        print("No script tags found!")
        sys.exit(1)
        
    js_code = scripts[0]
    
    with open('temp.js', 'w', encoding='utf-8') as f:
        f.write(js_code)
        
    print("JS code extracted to temp.js")
    
    # Use python's compile function to check basic syntax (it won't check JS syntax directly, but we can call node if available)
    # Let's check using node if present
    try:
        res = subprocess.run(['node', '--check', 'temp.js'], capture_output=True, text=True)
        if res.returncode == 0:
            print("Syntax OK (Node)")
        else:
            print("Syntax Error (Node):")
            print(res.stderr)
    except FileNotFoundError:
        print("Node not found, attempting general JS syntax parsing in Python...")
        # Simple bracket matcher
        stack = []
        lines = js_code.split('\n')
        for idx, line in enumerate(lines):
            for char in line:
                if char in '{[(':
                    stack.append((char, idx+1))
                elif char in '}])':
                    if not stack:
                        print(f"Extra closing character '{char}' at line {idx+1}")
                    else:
                        top, line_num = stack.pop()
                        if (char == '}' and top != '{') or (char == ']' and top != '[') or (char == ')' and top != '('):
                            print(f"Mismatched '{top}' (line {line_num}) and '{char}' (line {idx+1})")
        if stack:
            print("Unclosed brackets:")
            for char, line_num in stack[:5]:
                print(f"  '{char}' at line {line_num}")
                
except Exception as e:
    print("Execution failed:", e)
